File: src/lib/bindings/playlist.ts
import { invoke } from "@tauri-apps/api/core";

export interface PlaylistItem {
  index: number;
  filename: string;
  current: boolean;
  title: string;
}

export const playlistAdd = (path: string) => invoke<void>("playlist_add", { path });
export const playlistRemove = (index: number) => invoke<void>("playlist_remove", { index });
export const playlistNext = () => invoke<void>("playlist_next");
export const playlistPrev = () => invoke<void>("playlist_prev");
export const playlistPlayIndex = (index: number) => invoke<void>("playlist_play_index", { index });
export const playlistClear = () => invoke<void>("playlist_clear");
export const getPlaylist = () => invoke<PlaylistItem[]>("get_playlist");

File: src-tauri/src/commands/playlist.rs
use tauri::State;

use crate::error::AppError;
use crate::services::playlist::{PlaylistItem, PlaylistService};
use crate::state::MpvState;

#[tauri::command]
pub async fn playlist_add(path: String, state: State<'_, MpvState>) -> Result<(), AppError> {
    if !std::path::Path::new(&path).exists() {
        return Err(AppError::FileNotFound(path));
    }
    PlaylistService::add(state.get()?, &path)?;
    Ok(())
}

#[tauri::command]
pub async fn playlist_remove(index: i64, state: State<'_, MpvState>) -> Result<(), AppError> {
    PlaylistService::remove(state.get()?, index)?;
    Ok(())
}

#[tauri::command]
pub async fn playlist_next(state: State<'_, MpvState>) -> Result<(), AppError> {
    PlaylistService::next(state.get()?)?;
    Ok(())
}

#[tauri::command]
pub async fn playlist_prev(state: State<'_, MpvState>) -> Result<(), AppError> {
    PlaylistService::prev(state.get()?)?;
    Ok(())
}

#[tauri::command]
pub async fn playlist_play_index(index: i64, state: State<'_, MpvState>) -> Result<(), AppError> {
    PlaylistService::play_index(state.get()?, index)?;
    Ok(())
}

#[tauri::command]
pub async fn playlist_clear(state: State<'_, MpvState>) -> Result<(), AppError> {
    PlaylistService::clear(state.get()?)?;
    Ok(())
}

#[tauri::command]
pub async fn get_playlist(state: State<'_, MpvState>) -> Result<Vec<PlaylistItem>, AppError> {
    Ok(PlaylistService::get_all(state.get()?))
}
